flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java

```diff
@@ -139,6 +142,33 @@ public void testRegularExecution() throws Exception {
 		taskManagerActions.validateListenerMessage(ExecutionState.FINISHED, task, null);
 	}
 
+	@Test
+	public void testTaskTerminatesOnlyAfterInvokable() throws Exception {
```

Comment:
Is it covering for the user reported bug? It's not using `SourceStreamTask` anywhere, so it looks like it's not? If that's the case, couldn't we add a test in `SourceStreamTaskTest`?

Reply:
Added `SourceStreamTaskTest.testWatForCancelCompletion`.
For that, I extracted `SourceStreamTask.LegacySourceFunctionRunner` interface in a preceding hotfix commit.

Reply:
Hmm. 
> extracted SourceStreamTask.LegacySourceFunctionRunner interface in a preceding hotfix commit.

this commit is not super pretty, maybe we could make it a bit better, but I'm not sure if that's the right direction.

> SourceStreamTaskTest.testWatForCancelCompletion

this doesn't seem to be testing much, it's almost a no-op test for passing the returned value of one getter. It's definitely not testing for the reported bug.

What about creating a mock source that waits on a latch in a loop ignoring interrupts and the test thread fires cancellation & ensures task is not completed? Setup should be quite easy like in `SourceStreamTaskTest#finishingIgnoresExceptions`? What do you think?

Reply:
> It's definitely not testing for the reported bug.

A proper test for the reported bug would be an end-to-end test involving running a remote cluster, syncing source thread with class unloading and then checking the logs.
I think we agreed offline that such automated test would be an overkill and could rather be done manually. And the rest should be covered with unit tests. 

Now, this unit test checks that `SourceStreamTask` returns completion future of its thread. 
 This is what changed - and this is what I think **should** be tested.

If we also want to test `LegacySourceFunctionThread` future completion, it can be done independently now after extraction (however, I think this is out of the scope of this PR).

Furthermore, it's not possible to reliably test that a program does not perform some action if there is no time boundary (in this case test that task never completes if source thread never completes).

Reply:
Yes, e2e test would be an overkill, but semi integration test in `SourceStreamTask` seems to be quite easy.
>  it's not possible to reliably test that a program does not perform some action if there is no time boundary (in this case test that task never completes if source thread never completes).

What's the problem with waiting on the latch as I proposed?

Reply:
> What's the problem with waiting on the latch as I proposed?

For how long should the test thread wait on the latch (assuming `SourceStreamTask.sourceThread` never exits)?
Or did I misunderstood your proposal?

> It should provide automated test coverage for that bug.

It provides only partial coverage because it doesn't cover integration with `Task`, `StreamTask` and the code that actually unloads the classes. So in terms of coverage, it's equivalent to a set of unit tests.